### PR TITLE
fix backend check

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @muellerzr 
Related to #2652 . I think we'd better have a test on it.

I think the best way is to have an end-to-end test in the CI; it will avoid most issues : )

### CODE REVIEW

File: src/accelerate/state.py [hunk 1]
```diff
@@ -718,41 +721,45 @@ def _prepare_backend(
...
         ):
             if not cpu and is_xpu_available():
                 distributed_type = DistributedType.MULTI_XPU
             else:
                 distributed_type = DistributedType.MULTI_CPU
-            if is_ccl_available() and (
-                get_int_from_env(["CCL_WORKER_COUNT"], 0) > 0 or distributed_type == DistributedType.MULTI_XPU
+
+            if (
+                (backend is None or backend == "ccl")
```
Comment (muellerzr) [inline src/accelerate/state.py:746]:
```suggestion
                backend in (None, "ccl")
```

File: src/accelerate/state.py [hunk 2]
```diff
@@ -718,41 +721,45 @@ def _prepare_backend(
...
+                and (get_int_from_env(["CCL_WORKER_COUNT"], 0) > 0 or distributed_type == DistributedType.MULTI_XPU)
             ):
                 if get_ccl_version() >= "1.12":
                     import oneccl_bindings_for_pytorch  # noqa: F401
                 else:
                     import torch_ccl  # noqa: F401
 
                 backend = "ccl"
-            elif torch.distributed.is_mpi_available():
+            elif (backend is None or backend == "mpi") and torch.distributed.is_mpi_available():
```
Comment (muellerzr) [inline src/accelerate/state.py:756]:
```suggestion
            elif (backend in (None, "mpi")) and torch.distributed.is_mpi_available():
```

File: src/accelerate/state.py [hunk 3]
```diff
@@ -179,7 +179,10 @@ def __init__(self, cpu: bool = False, **kwargs):
                 )
 
             # Sets up self.backend + imports
-            backend, distributed_type = self._prepare_backend(cpu, use_sagemaker_dp, kwargs.pop("backend", None))
+            original_backend = kwargs.pop("backend", None)
+            backend, distributed_type = self._prepare_backend(cpu, use_sagemaker_dp, original_backend)
+            if original_backend and backend != original_backend:
+                logger.warning(f"The assigned backend is {original_backend}, but the real backend is {backend}")
```
Comment (muellerzr) [inline src/accelerate/state.py:185]: I don't think it makes sense to do a warning here. This should be a `raise` if this is truly a situation we can get into.

**Review by muellerzr — COMMENTED** [review]
Thanks! This looks a bit better on paper at least, I'll pull it down and test locally before giving the ✅ 

Left a few logic and style nits

File: src/accelerate/state.py (hunk 3 above)
Comment (jiqing-feng) [reply src/accelerate/state.py:185]: Done

Comment (muellerzr) [reply src/accelerate/state.py:185]: I don't see this change propagated?

Comment (jiqing-feng) [reply src/accelerate/state.py]: Sorry, I forgot to push. It should be updated now, pls check. Thx.

File: src/accelerate/state.py [hunk 4]
```diff
@@ -179,7 +179,10 @@ def __init__(self, cpu: bool = False, **kwargs):
                 )
 
             # Sets up self.backend + imports
-            backend, distributed_type = self._prepare_backend(cpu, use_sagemaker_dp, kwargs.pop("backend", None))
+            original_backend = kwargs.pop("backend", None)
+            backend, distributed_type = self._prepare_backend(cpu, use_sagemaker_dp, original_backend)
+            if original_backend and backend != original_backend:
```
Comment (muellerzr) [inline src/accelerate/state.py:184]: One last final check, sorry missed this the first time. Let's not trust python here 😉 
```suggestion
            if original_backend is not None and backend != original_backend:
```

**Review by muellerzr — APPROVED** [review]
Thanks so much! One last, last thing then we can merge and get a patch out ✅

File: src/accelerate/state.py (hunk 4 above)
Comment (jiqing-feng) [reply src/accelerate/state.py]: Done.

**Review by muellerzr — APPROVED** [review]
Thanks for fixing!

**Pull request merged**